### PR TITLE
feat: Add tones support for the RPi Pico

### LINKED ISSUE #1230: Issue with APRS using AFSK on the RPi Pico
Hi there,

I am trying to transmit APRS packets using AFSK on the RPi Pico (RP2040) and SX1278 module. I am starting from the HAL provided in the examples/NonArduino/Pico. 

My first step was to override the `tone` and `noTone` functions so that they can generate square waves at given frequencies:

```cpp
// in PicoHal.h
void set_pwm_frequency(uint gpio, unsigned int frequency) {
    uint slice_num = pwm_gpio_to_slice_num(gpio);

    // Calculate the PWM wrap value based on the desired frequency
    pwm_config config = pwm_get_default_config();
    uint32_t clock_freq = clock_get_hz(clk_sys);

    float divider = clock_freq / (frequency * 65536.0);
    pwm_config_set_clkdiv(&config, divider);
    pwm_init(slice_num, &config, true);

    uint32_t wrap = clock_freq / frequency;

    uint16_t level = (1 << 15);
    pwm_set_gpio_level(gpio, level);
}

void tone(uint32_t pin, unsigned int frequency, unsigned long duration = 0) override {
      printf("Using pin %d at freq %d...", pin, frequency);
      gpio_set_function(pin, GPIO_FUNC_PWM);
      set_pwm_frequency(pin, frequency);

      if (duration > 0){
              sleep_ms(duration);
              pwm_set_enabled(pwm_gpio_to_slice_num(pin), false);
      }
}

void noTone(uint32_t pin) override {
      pwm_set_enabled(pwm_gpio_to_slice_num(pin), false);
}
```

I have tried the AFSK_Tone and AFSK_Imperial_March examples, and they work fine with this implementation. However, when I try to transmit APRS using AFSK, it sounds like its just using FSK: like it's not using my tone generators at all. Here's my code for that"

```cpp
// define pins to be used
#define SPI_PORT spi1
#define SPI_MISO 12
#define SPI_MOSI 11
#define SPI_SCK 10

#define RFM_NSS 13
#define RFM_RST 22
#define RFM_DIO0 14
#define RFM_DIO1 15

#include <pico/stdlib.h>
#include <cstdlib>
// include the library
#include <RadioLib.h>
#include "melody.h"
#include "hardware/pwm.h"

// include the hardware abstraction layer
#include "PicoHal.h"

PicoHal* hal = new PicoHal(SPI_PORT, SPI_MISO, SPI_MOSI, SPI_SCK);
SX1278 radio = new Module(hal, RFM_NSS, RFM_DIO0, RFM_RST, RFM_DIO1);

AFSKClient audio(&radio, 28);
AX25Client ax25(&audio);
APRSClient aprs(&ax25);

int main() {
  stdio_init_all();

  printf("[SX1278] Initializing ... ");
  int state = radio.beginFSK(434.0, 48, 2.2);
  if (state != RADIOLIB_ERR_NONE) {
    printf("failed, code %d\n", state);
    return(1);
  }

  state = ax25.begin("MXLMX");
  if (state != RADIOLIB_ERR_NONE) {
    printf("failed, code %d\n", state);
    return(1);
  }

  aprs.begin('>');

  // loop forever
  for(;;) {
    // send a packet
    printf("[SX1278] Transmitting tone ... ");
    int state = aprs.sendPosition("N0CALL", 0, "4911.67N", "01635.96E");
    
    sleep_ms(3000);
  }

  return(0);
}
```

I have tried different configurations of the SX1278, to no avail. I have checked the state at radio initialization, ax25 initialization, and aprs initialization. Nothing wrong there. This is what a waterfall plot of what I'm seeing looks like:

image

I know this library is targeted to more Arduino users. However, I feel like I am at the finish line with getting this to work on the RPi Pico and SX1278. Is there anything glaring I'm missing? I'll include the full HAL file as a comment to this issue so it doesn't get too unwieldy. Thanks!

### PR DESCRIPTION
**Pull request by mosa11aei**
The following PR adds overrides for the `tone()` and `noTone()` functions in the HAL for the RPi Pico. Tones are added based on bitbangs.

This code was tested using the APRS_Position.ino example, modified for the Pico. Closes #1230

### CODE REVIEW

**Comment by jgromes** [comment]
Looks like the CI for rpi-pico cannot build it. Which version of the SDK are you using? The CI checks out 1.5.1.

**Comment by mosa11aei** [comment]
This isn't a problem with the SDK actually. I forgot to commit my changes to the CMakeLists.txt file, my apologies. Let me do that quickly.

**Comment by mosa11aei** [comment]
@jgromes it should build properly now!

File: examples/NonArduino/Pico/main.cpp [hunk 1]
```diff
@@ -44,7 +44,7 @@
 #include "PicoHal.h"
 
 // create a new instance of the HAL class
-PicoHal* hal = new PicoHal(SPI_PORT, SPI_MISO, SPI_MOSI, SPI_SCK);
+PicoHal* hal = new PicoHal(SPI_PORT, SPI_MISO, SPI_MOSI, SPI_SCK, RADIOLIB_NC);
```
Comment (jgromes) [inline examples/NonArduino/Pico/main.cpp:47]: The need for this new PWM pin is not really explained here, and I'm not sure it is necessary. The only thing it's sued for is configuring it as output, which can be done in the main file. Most users will set this to `RADIOLIB_NC` anyway.

File: examples/NonArduino/Pico/PicoHal.h [hunk 2]
```diff
@@ -110,6 +151,19 @@ class PicoHal : public RadioLibHal {
...
+    toneLoopPin = pin;
+    toneLoopFrequency = frequency;
+    toneLoopDuration = duration;
+    multicore_launch_core1(toneLoop);
+  }
+
+  void noTone(uint32_t pin) override {
+    multicore_reset_core1();
+  }
+
```
Comment (jgromes) [inline examples/NonArduino/Pico/PicoHal.h:166]: Interesting imlementation, to offload this to the other core. However, I would add a large warning at the beginning of this file that this is being done, as I suspect most people will not expect it.

**Comment by jgromes** [comment]
Indeed, it does build now. I also left two minor comments, after those are addressed I think this can be merged.

File: examples/NonArduino/Pico/main.cpp (hunk 1 above)
Comment (mosa11aei) [reply examples/NonArduino/Pico/main.cpp:47]: I can get behind that logic. That will be removed.

File: examples/NonArduino/Pico/PicoHal.h (hunk 2 above)
Comment (mosa11aei) [reply examples/NonArduino/Pico/PicoHal.h:166]: Actually, you can't run this process on the one core if you're trying to do other tasks on it. There is a lot of jitter in the output that is caused. Thus, this process is offloaded to the other core, and jitter becomes unnoticeable and transmissions become much more consistent. Comment will be added.

**Review by jgromes — APPROVED** [review]

**Comment by jgromes** [comment]
All looks good now - merged, thank you for the contribution!

**Pull request merged**